compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -250,6 +301,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
         let get_abi = |this: &Self, instance_ty: Ty<'tcx>| match instance_ty.kind() {
             ty::FnDef(..) => instance_ty.fn_sig(*this.tcx).abi(),
+            // Even after lowering closures and generators away, the *callee* can still have this
+            // kind of type.
```

Comment:
Using `instance_ty.fn_sig(*this.tcx)` in all cases works just fine for cg_clif after monomorphization: https://github.com/bjorn3/rustc_codegen_cranelift/blob/a65a523d800b1937bf0f6e9857015d1a1c585a62/src/abi/mod.rs#L316

Reply:
> Yes but I don't think codegen backends call this on the instance obtained from the MIR body?

It is obtained from the `Call` terminator in the linked cg_clif code. This function is called from https://github.com/bjorn3/rustc_codegen_cranelift/blob/122219237437ee1deee33df9806a4316194a6f76/src/base.rs#L360-L371

Reply:
Yes, that is what Miri calls the *caller* instance. The `match` here used to be called on `self.tcx.type_of(callee_def_id)`, which returns a different type. (Now it is called on the instance returned by `find_mir_or_eval_fn` so possibly `fn_sig` could work now, I haven't checked again since doing that change.)

But anyway I will remove `get_abi` entirely so it doesn't really matter.

Reply:
A long comment was removed, so I am confused now... but the PR could only ever lead to new things being accepted `check_argument_compat` for `!rust_abi` where previously we'd require full type equality and now we instead require mode equality.

Reply:
> A long comment was removed, so I am confused now...

It is one discussion thread up at https://github.com/rust-lang/rust/pull/91342#discussion_r760454076.

> but the PR could only ever lead to new things being accepted in check_argument_compat for !rust_abi

It also leads to more code being accepted for `rust_abi` where the indirect and cast pass modes were previously always considered incompatible. In the rust abi it is unstable if an argument will be passed as indirect or cast. I can't think of any valid reasons to allow mismatched types for them even for libstd internal usage.